storage/replica_leader_lease.go

```diff
@@ -20,100 +20,164 @@
-			// send it earlier the race can still happen).
-			if i == len(r.mu.llChans)-1 {
+		replica.mu.Lock()
+		defer replica.mu.Unlock()
+		for i, llChan := range p.llChans {
+			// Don't send the same pErr object twice; this can lead to races. We could
+			// clone every time but it's more efficient to send pErr itself to one of
+			// the channels (the last one; if we send it earlier the race can still
+			// happen).
+			if i == len(p.llChans)-1 {
 				llChan <- pErr
 			} else {
 				llChan <- protoutil.Clone(pErr).(*roachpb.Error) // works with `nil`
 			}
 		}
-		r.mu.llChans = r.mu.llChans[:0]
+		p.llChans = p.llChans[:0]
+		p.nextLease = roachpb.Lease{}
 	}) {
 		// We failed to start the asynchronous task. Send a blank NotLeaderError
 		// back to indicate that we have no idea who the leader might be; we've
 		// withdrawn from active duty.
-		llChan <- roachpb.NewError(r.newNotLeaderErrorLocked(nil, r.store.StoreID()))
+		llChan <- roachpb.NewError(
+			replica.newNotLeaderError(nil, replica.store.StoreID(), replica.mu.desc))
 		return llChan
 	}
-
-	r.mu.llChans = append(r.mu.llChans, llChan)
+	p.llChans = append(p.llChans, llChan)
+	p.nextLease = leaseReq.Lease
 	return llChan
 }
+
+func (r *Replica) getOrExtendLeaderLeaseLocked(timestamp roachpb.Timestamp) <-chan *roachpb.Error {
+	// Propose a Raft command to get a leader lease for this replica.
+	desc := r.mu.desc
+	_, replica := desc.FindReplica(r.store.StoreID())
+	if replica == nil {
+		llChan := make(chan *roachpb.Error, 1)
+		llChan <- roachpb.NewError(roachpb.NewRangeNotFoundError(r.RangeID))
+		return llChan
+	}
+	if r.store.IsDrainingLeadership() {
+		// We've retired from active duty.
+		llChan := make(chan *roachpb.Error, 1)
+		llChan <- roachpb.NewError(r.newNotLeaderError(nil, r.store.StoreID(), r.mu.desc))
+		return llChan
+	}
+	return r.mu.pendingLeaderLeaseRequest.InitOrJoinRequest(
+		r, *replica, timestamp, desc.StartKey.AsRawKey())
+}
+
+// executeLeaderLeaseCommand proposes a LeaderLeader and waits for its
+// completion.
+func (r *Replica) executeLeaderLeaseCommand(
+	nextLease roachpb.LeaderLeaseRequest,
+	timestamp roachpb.Timestamp,
+	startKey roachpb.Key,
+) *roachpb.Error {
+	ba := roachpb.BatchRequest{}
+	ba.Timestamp = r.store.Clock().Now()
+	ba.RangeID = r.RangeID
+	ba.Add(&nextLease)
+
+	// Send lease request directly to raft in order to skip unnecessary
+	// checks from normal request machinery, (e.g. the command queue).
+	// Note that the command itself isn't traced, but usually the caller
+	// waiting for the result has an active Trace.
+	ch, _, err := r.proposeRaftCommand(r.context(context.Background()), ba)
+	if err != nil {
+		return roachpb.NewError(err)
+	}
+
+	// If the command was committed, wait for the range to apply it.
+	select {
+	case c := <-ch:
+		if c.Err != nil {
+			if log.V(1) {
+				log.Infof("failed to acquire leader lease for replica %s: %s", r.store, c.Err)
+			}
+		}
+		return c.Err
+	case <-r.store.Stopper().ShouldDrain():
```

Comment:
Add a comment on l.166 that once we use a real context there, we can't do this if we don't manage to abandon the command (see `addWriteCmd`).